### PR TITLE
Uuid: add default ctor and Parse()

### PR DESCRIPTION
**Pull request by antkmsft**
Interestingly, we could even make `Parse()` a `constexpr`, but I do not want the exception message to live in the header, so that the string message we pass to it does not live in multiple translation units.

`ToString()` can also be made `constexpr`, for the stdlib implementations where std::string constructors are constexpr (not until C++20?). But still, with some template magic, if we really wanted to, we could make it conditionally constexpr, where available.

I did not make either, sounds like an overkill, unless we really-really needed it.
But for other methods, I think we could make them constexpr without sacrificing much, so I made them constexpr.

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 1]
```diff
@@ -54,6 +58,46 @@ namespace Azure { namespace Core {
...
+    /**
+     * @brief Construct a Uuid by parsing its representation.
+     * @param s a string in `8-4-4-4-12` hex characters format.
+     * @throw `std::invalid_argument` if \p s cannot be parsed.
+     */
+    static Uuid Parse(std::string const& s);
+
+    constexpr bool operator==(Uuid const& other) const
+    {
+      // std::array::operator==() is not a constexpr until C++20
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:72]: if (m_uuid.size() != other.m_uuid.size())
   return false;

File: sdk/core/azure-core/src/uuid.cpp [hunk 2]
```diff
@@ -83,19 +96,77 @@ namespace Azure { namespace Core {
...
+        {
+          if (c != '-')
+          {
+            parseError = true;
+            break;
+          }
+        }
+        else
+        {
+          const auto c2 = s[i + 1];
```
Comment (LarryOsterman) [inline sdk/core/azure-core/src/uuid.cpp]: Add an assert that i+1 is less than UuidStringLength? Otherwise I think this might run off the end of the string (the loop invariant is on i < UuidStringLength so I believe this canaccess UuidStringLength into the string, which is 1 past the end?

File: sdk/core/azure-core/src/uuid.cpp [hunk 3]
```diff
@@ -18,61 +20,135 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
+               + "' is not in the '00112233-4455-6677-8899-aAbBcCdDeEfF' format.")
+                      : result;
+  }
+
   Uuid Uuid::CreateUuid()
   {
-    uint8_t uuid[UuidSize] = {};
+    Uuid result{};
+
+    using RngResultType = std::uint32_t;
```
Comment (RickWinter) [inline sdk/core/azure-core/src/uuid.cpp:130]: Can you explain why you do the using. Is it to ensure the same type is used throughout the code? If so I'd add a comment for why RngResultType exists.

**Review by LarryOsterman — APPROVED** [review]

File: sdk/core/azure-core/test/ut/uuid_test.cpp [hunk 4]
```diff
@@ -127,3 +127,49 @@ TEST(Uuid, validChars)
...
+           0xAA,
+           0xBB,
+           0xCC,
+           0xDD,
+           0xEE,
+           0xFF}));
+
+  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000-000000000000 "), std::invalid_argument);
+  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000-00000000000"), std::invalid_argument);
+  ASSERT_THROW(Uuid::Parse("00000000a0000-0000-0000-000000000000"), std::invalid_argument);
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/uuid_test.cpp:169]: Add variations for all the edge cases. Also would be good to test upper case, lower case, symbols, white space ....

```suggestion
  ASSERT_THROW(Uuid::Parse("00000000a0000-0000-0000-000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000a0000-0000-000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000-0000a0000-000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000a000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000000000000000000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-000000000000000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000-00000000000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000000000000000"), std::invalid_argument);
  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000-0000000000G"), std::invalid_argument);
```

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 1 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:72]: I don't see how it is realistically possible - std::array has a constexpr size as it's template parameter.

File: sdk/core/azure-core/src/uuid.cpp (hunk 3 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:130]: Yes, that's exactly why - to highlight that "the exact same type" is used in several places. I'll add a comment.

File: sdk/core/azure-core/src/uuid.cpp [hunk 5]
```diff
@@ -83,19 +96,77 @@ namespace Azure { namespace Core {
     // values are reserved for backward compatibility. The C|c, D|d values are reserved for
     // Microsoft, and the E|e, F|f values are reserved for future use.
     // Therefore, we have to zero out the two high bits, and then set the highest bit to 1.
-    uuid[8] = (uuid[8] & 0x3F) | 0x80;
+    result.m_uuid.data()[8] = (result.m_uuid.data()[8] & 0x3F) | 0x80;
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:162]: Good idea, I'll add it. Thank you!

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 4 above)
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:169]: Ok, I'll add more - for whitespace and random characters, for example. As far as upper/lowercase, there is a test (parsing `"00112233-4455-6677-8899-aAbBcCdDeEfF"`).

File: sdk/core/azure-core/src/uuid.cpp (hunk 2 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp]: Yes, in theory it could, I see, this could be more fragile, if a developer in the future would modify this code incorrectly, I don't see how it wouldn't be caught by the tests though, but I'll see what I can do, I think I'll end up adding the AZURE_ASSERT as you're suggesting.

Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp]: I ended up adding `assert()`. I saw you are also using it. `AZURE_ASSERT()` seems like an overkill to me, because it does compile in release configurations as well, and aborts if condition is not true, which might be right in some circumstances, but I don't think it should be used here, especially that it is used inside a loop.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 6]
```diff
@@ -8,32 +8,40 @@
...
   class Uuid final {
+  public:
+    /**
+     * @brief Represents a byte array where the UUID value can be stored.
+     *
+     */
+    using ValueArray = std::array<std::uint8_t, 16>;
 
   private:
-    static constexpr size_t UuidSize = 16;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:26]: Why did we remove this named constant?

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 7]
```diff
@@ -42,7 +50,7 @@ namespace Azure { namespace Core {
      * representation of the Uuid.
      * @returns An array with the binary representation of the Uuid.
      */
-    std::array<uint8_t, UuidSize> const& AsArray() const { return m_uuid; }
+    constexpr ValueArray const& AsArray() const { return m_uuid; }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:53]: Just to confirm, changing to an alias like this is never a breaking change?

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 8]
```diff
@@ -54,6 +62,60 @@ namespace Azure { namespace Core {
...
      */
-    static Uuid CreateFromArray(std::array<uint8_t, UuidSize> const& uuid);
+    static constexpr Uuid CreateFromArray(ValueArray const& uuid) { return Uuid{uuid}; }
+
+    /**
+     * @brief Construct a Uuid by parsing its representation.
+     * @param s a string in `8-4-4-4-12` hex characters format.
+     * @throw `std::invalid_argument` if \p s cannot be parsed.
+     */
+    static Uuid Parse(std::string const& s);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:72]: We should avoid single letters/abbreviated parameter names that are not descriptive.
`uuidString`?

File: sdk/core/azure-core/CHANGELOG.md [hunk 9]
```diff
@@ -5,11 +5,14 @@
 ### Features Added
 
 - Request logs to now include the `accept-range`, `content-range`, `range`, `WWW-Authenticate`, `x-ms-date`, `x-ms-error-code`, `x-ms-range`, and `x-ms-version` headers.
+- Added default constructor, `Parse()`, and equality comparison operators to `Azure::Core::Uuid`.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:8]: Who do we expect to use the Parse method today (in the SDK/end user), or are we adding it preemptively for a near-future/known use case?

File: sdk/core/azure-core/inc/azure/core/uuid.hpp [hunk 10]
```diff
@@ -54,6 +62,60 @@ namespace Azure { namespace Core {
...
+     * @return `true` if values of two Uuids are not equal, `false` otherwise.
+     *
+     */
+    constexpr bool operator!=(Uuid const& other) const { return !(*this == other); }
+
+    /**
+     * @brief Checks if the value represents a Nil UUID (`00000000-0000-0000-0000-000000000000`).
+     *
+     */
+    constexpr bool IsNil() const
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/uuid.hpp:106]: In which scenario is this a useful helper?

Should we rather have a static 'Nil' or 'Zero' uuids constant, and then we don't need this helper method. Instead, the caller can use the equality operators for this check, instead.

Something like:
```c++
if (myUuid == Uuids::ZeroUuid)
{
// Take some action
}
```

This way, we don't need this additional method.

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:106]: Similar to .NET's "Empty"
https://learn.microsoft.com/en-us/dotnet/api/system.guid.empty?view=net-8.0#system-guid-empty

That is a more descriptive and unsurprising design, rather than the default ctor resulting in a "sentinel" nil value. I don't think having a default ctor on Uuid conveys the right intent.

Unless it is absolutely necessary, I would rather not have it in the public surface area. It doesn't have a "real" semantic meaning, on its own.

File: sdk/core/azure-core/src/uuid.cpp [hunk 11]
```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
+  if (c >= 'A')
+  {
+    return 10 + (c - 'A');
+  }
 
-  return 'a' + (byte - 10);
+  return c - '0';
+}
+
+constexpr char NibbleToHex(std::uint8_t nibble) // does not check for errors
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:54]: Was it necessary to remove the azure_assert to make it a constexpr?
If not, than we should keep that invariant check in code, rather than a comment.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/uuid.cpp:54]: And do we need it to be a constexpr?

File: sdk/core/azure-core/src/uuid.cpp [hunk 12]
```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
 } // namespace
 
 namespace Azure { namespace Core {
   std::string Uuid::ToString() const
   {
-    std::string s(36, '-');
+    std::string s(UuidStringLength, '-');
 
-    for (size_t i = 0, j = 0; j < s.size() && i < UuidSize; i++)
+    for (size_t bi = 0, si = 0; bi < m_uuid.size(); ++bi)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:70]: What's si and bi? StringIndex and ByteIndex? Probably worth expanding.

File: sdk/core/azure-core/src/uuid.cpp [hunk 13]
```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
-      uint8_t highNibble = (m_uuid[i] >> 4) & 0x0F;
-      uint8_t lowNibble = m_uuid[i] & 0x0F;
-      s[j++] = ByteToHexChar(highNibble);
-      s[j++] = ByteToHexChar(lowNibble);
+      assert((si < UuidStringLength) && (si + 1 < UuidStringLength));
+
+      const std::uint8_t b = m_uuid[bi];
+      s[si] = NibbleToHex((b >> 4) & 0x0F);
+      s[si + 1] = NibbleToHex(b & 0x0F);
+      si += 2;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:82]: We changed the few lines here, subtly.
I am curious, why?
For example:
Previously we incremented j++, now we don't and at the end do si += 2

Is that better?
And why remove the named locals for the nibbles, rather than pass the transformed value directly in the method call?

Trying to see the benefit. Imo, neither of those changes help readability and maintainability.

File: sdk/core/azure-core/src/uuid.cpp [hunk 14]
```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
+  }
+
   Uuid Uuid::CreateUuid()
   {
-    uint8_t uuid[UuidSize] = {};
+    Uuid result{};
+
+    // Using RngResultType and RngResultSize to highlight the places where the same type is used.
+    using RngResultType = std::uint32_t;
+    static_assert(sizeof(RngResultType) == 4, "sizeof(RngResultType) must be 4.");
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:139]: Tbh, introducing these two "types" makes the rest of the implementation mor complicate to read and parse.

I don't think it's worth it here. Aliasing uint32 in particular adds more confusion than the value we are trying to provide here.

File: sdk/core/azure-core/src/uuid.cpp [hunk 15]
```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
...
-
-    return Uuid(uuid);
-  }
+    {
+      // https://www.rfc-editor.org/rfc/rfc9562.html#name-version-field
+      constexpr std::uint8_t Version = 4; // Version 4: Pseudo-random number
+      result.m_uuid.data()[6] = (result.m_uuid.data()[6] & 0xF) | (Version << 4);
+    }
 
-  Uuid Uuid::CreateFromArray(std::array<uint8_t, UuidSize> const& uuid)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:95]: Why move the impl to the header file? Trying to understand the benefit of either approach. Do we prefer implementation to be in cpp file over hpp file or vice versa?

File: sdk/core/azure-core/test/ut/uuid_test.cpp [hunk 16]
```diff
@@ -127,3 +127,81 @@ TEST(Uuid, validChars)
...
+  // Empty string
+  ASSERT_THROW(Uuid::Parse(""), std::invalid_argument);
+
+  // Special characters - make sure we're not treating them as byte array
+  ASSERT_THROW(Uuid::Parse("\a\a\a\a\a\a\a\a\a\a\a\a\a\a\a\a"), std::invalid_argument);
+
+  // Spaces before, after, and both.
+  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000-000000000000 "), std::invalid_argument);
+  ASSERT_THROW(Uuid::Parse(" 00000000-0000-0000-0000-000000000000"), std::invalid_argument);
+  ASSERT_THROW(Uuid::Parse("00000000-0000-0000-0000-00000000000"), std::invalid_argument);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/uuid_test.cpp:176]: I don't see spaces before/after in this test case. Comment/test mismatch?

I am assuming we expect a "zero" Uuid parse to fail?

File: sdk/core/azure-core/src/uuid.cpp (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/uuid.cpp:54]: Looking at your other comments:
> AZURE_ASSERT() seems like an overkill to me, because it does compile in release configurations as well,...

I see. That makes sense to not use in loops, to enforce a variant we know/expect to be true.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 6 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:26]: It is not used anywhere anymore. In place of it, where it is needed, we use `m_uuid.size()`, which is a constexpr, so effectively the same constant.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 7 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:53]: No, it is not.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 10 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:106]: I don't feel strongly about it bein either way, but Nil is an RFC's concept - not "Empty", not "Zero" - it is called Nil.
I think `IsNil()` is ok, it is similar to `HasValue()` for `Nullable`. It is possible to write a condition `if (!x.IsNil()) { /* serialize it */}`

`x == Uuid{}`, or `x == Uuid::Nil` are a bit more expensive. I don't think it will ever be significant though.

`x == Uuid{}` is a bit less readable, because you have to understand that `Uuid{}` produces Nil UUD.
But it is still possible to write it, if user wants it.

`Uuid::Nil` would require a DLLEXPORT, it is more expensive than `IsNil()`, it won't be a constexpr, while it also adds an API. So, we chose, which API to add, `IsNil()`, or `Uuid::Nil`, and I like `IsNil()` more.

File: sdk/core/azure-core/src/uuid.cpp (hunk 11 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:54]: Yes, that is for the loop reason. Everything at the moment is pretty scoped here, plus it is covered with unit tests, so I think making the function to check for errors in a loop would be a bit overkill, not to mention the fact that it will be there for release builds as well.

As for the constexpr - it does not have to be constexpr, but if it is constexpr, it is better - it allows the compiler to do more optimizations, potentially producing more effective code.

File: sdk/core/azure-core/src/uuid.cpp (hunk 13 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:82]: It is pretty equal.
I didn't feel the need for there to be a local variable that is only used once, and I don't think it is hard to understand what it is used for, without making it a dedicated variable. It adds two lines (three if you add a space after), so function would take more lines on the screen increasing the mental resources you need to read it. Plus, previously the nobble variables were not `const`, so you as a reader might question if they get changed anywhere (does `ByteToHexChar()` take `uint8_t&` and modify them?), or if they are used in two places.

One `x+=2;` is potentially more effective than `x++; x++;` and even `++x; ++x;`, plus it makes an expression without side-effects (we're both assigning and incrementing on the same line; historically, this can sometimes bite you, so I avoid writing expressions with side effects just as a habit).

File: sdk/core/azure-core/src/uuid.cpp (hunk 14 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:139]: But if what we use in place of `RngResultType` is not effectively a `std::uint32_t` on some implementation, and if it's size is not `4`, it may behave incorrectly, and not be caught for a long time.
Here, we have a `static_assert` ensuring that our assumptions are correct. If it ever explodes, you need to review/update all the lines below that use `RngResultType` or `RngResultSize`. That is why they are dedicated named constants, because they express our assumptions and you don't have to review any of the lines that do not involve these two constants, if the `static_assert` explodes.

File: sdk/core/azure-core/src/uuid.cpp (hunk 15 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:95]: Having it in `.hpp` file allows us to make it `constexpr`, because `std::array` initialization from another `std::array` is a constexpr, and is trivial in terms of assembly commands. The implementation is not multiple lines long, so should it be just inlined, the code size impact is tiny, probably the same number of assembler commands as function call, if not less. And also it does not require any extra includes to be made in the `.hpp` file, so compilation cost wise it is pretty lightweight as well.

I know that we are sending stuff over the network, so hardly any place would ever be a bottleneck in our SDK, yet I think it makes sense to not pessimize implementation, especially if it comes basically free and we can use it easily.
I mean, the other day, you would make a PR optimizing the `Uuid::ToString()` performance, and I and the team liked the change and agreed to it (caching was another thing, chacing would have tradeoffs, while writing simple parser instead of `sscanf()` did not have tradeoffs).

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 16 above)
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: Thanks! See #5943.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 8 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:72]: Thanks! See #5943.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 10 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:106]: > `Uuid::Nil` would require a DLLEXPORT, it is more expensive than `IsNil()`

What do you mean by, it is more expensive?

File: sdk/core/azure-core/src/uuid.cpp (hunk 13 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/uuid.cpp:82]: When I initially saw the code, without named locals, it wasn't clear what the bitwise operation/expressions were doing.

Even though the locals were used only once, they described the semantic meaning. Now, a few months later, we'll have to re-page back and understand what these are, and which one is the high and low nibble.

I would recommend keeping the named locals (or at least add them as a comment next each bitwise expression line).

The compiler will do the right thing here, in any case, so expressiveness/readability for humans takes the win here.

File: sdk/core/azure-core/src/uuid.cpp (hunk 14 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/uuid.cpp:139]: When is it not 4? We have had this code as is for quite a while now, and have heard no issues  on any compiler/OS.

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 16 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: Btw, isn't it surprising that parsing a nil Uuid fails? Even though we generate it, in the default ctor?

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 10 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:106]: Compiler may treat it as a variable, not making an assumptions about its value at compile-time. Comparison with another array is a byte-by-byte comparison, `cmp *ptr, *ptr`, while `IsNil()` is `cmp *ptr, 0` or even `xor *ptr, *ptr`/`sub *ptr, *ptr`.

File: sdk/core/azure-core/src/uuid.cpp (hunk 13 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:82]: I have added more comments, and the follow-up PR I created is already merged, do you think I should open another for this? I would prefer, as you are back, if you made a PR and update it as you see fits the best, I think I would approve it (my only suggestion is to declare such variables `const`). It should be ok to update it later, as first this would likely be before the release, second it should be ok to update even after release.

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 16 above)
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: That one had one less zero in the last section. I have moved it into a section of its own, and added 4 more possible variants.

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 6 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:26]: I see `16` used now, as an unnamed constant.

File: sdk/core/azure-core/src/uuid.cpp (hunk 13 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/uuid.cpp:82]: Yeah, sounds good.
I can address it once I am back, with some other changes.
Thanks!

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 16 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: Okay, great!

I think it's worth having a test for Parse(default ctor.ToString()).

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: I see you added that scenario (in spirit) already. Thanks

File: sdk/core/azure-core/inc/azure/core/uuid.hpp (hunk 6 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/uuid.hpp:26]: Yes, but that should be the only instance of `16` being used. The least should be replaced with `m_uuid.size()` which is effectively a constexpr 16.

File: sdk/core/azure-core/test/ut/uuid_test.cpp (hunk 16 above)
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/uuid_test.cpp:176]: Yes, we are testing that behavior, maybe not as explicitly, but it is covered. In your future PR, in your future PR, feel free to add a test for `Parse(Uuid{}.ToString())`.

File: sdk/core/azure-core/src/uuid.cpp (hunk 14 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/uuid.cpp:139]: We don't know if it has never been 4, most likely not, but now we have a `static_assert` which will fail to compile if it is not 4.
